File: kotlin/src/commonMain/kotlin/com/actualbudget/sync/db/SyncRepository.kt
package com.actualbudget.sync.db

import com.actualbudget.sync.proto.Message
import com.actualbudget.sync.proto.MessageEnvelope

/**
 * Repository for syncing CRDT messages to the local database.
 */
class SyncRepository(private val db: ActualDatabase) {

    /**
     * Apply a list of message envelopes to the database.
     * Messages are stored in the CRDT log and applied to their respective tables.
     */
    fun applyMessages(envelopes: List<MessageEnvelope>) {
        envelopes.forEach { envelope ->
            if (!envelope.isEncrypted) {
                try {
                    val message = envelope.decodeMessage()
                    applyMessage(envelope.timestamp, message)
                } catch (e: Exception) {
                    // Log error but continue processing
                    println("Failed to apply message ${envelope.timestamp}: ${e.message}")
                }
            }
        }
    }

    /**
     * Apply a single CRDT message.
     */
    private fun applyMessage(timestamp: String, message: Message) {
        // Store in CRDT log (value stored as BLOB)
        db.actualDatabaseQueries.insertMessage(
            timestamp = timestamp,
            dataset = message.dataset,
            row = message.row,
            column = message.column,
            value_ = message.value.encodeToByteArray()
        )

        // Parse the value
        val parsedValue = parseValue(message.value)

        // Ensure row exists, then update the column
        when (message.dataset) {
            "accounts" -> {
                ensureAccountExists(message.row)
                applyAccountColumn(message.row, message.column, parsedValue)
            }
            "payees" -> {
                ensurePayeeExists(message.row)
                applyPayeeColumn(message.row, message.column, parsedValue)
            }
            "payee_mapping" -> {
                ensurePayeeMappingExists(message.row)
                applyPayeeMappingColumn(message.row, message.column, parsedValue)
            }
            "categories" -> {
                ensureCategoryExists(message.row)
                applyCategoryColumn(message.row, message.column, parsedValue)
            }
            "category_groups" -> {
                ensureCategoryGroupExists(message.row)
                applyCategoryGroupColumn(message.row, message.column, parsedValue)
            }
            "transactions" -> {
                ensureTransactionExists(message.row)
                applyTransactionColumn(message.row, message.column, parsedValue)
            }
            "zero_budgets" -> {
                ensureBudgetExists(message.row)
                applyBudgetColumn(message.row, message.column, parsedValue)
            }
            "zero_budget_months" -> {
                // Buffer amounts per month - not critical for basic budget display
                // Can be implemented later for carryover feature
            }
            "schedules" -> {
                // Handle schedules
            }
            "rules" -> {
                // Handle rules
            }
            "notes" -> {
                // Handle notes
            }
            else -> {
                // Unknown dataset, skip
            }
        }
    }

    /**
     * Parse Actual's typed value format.
     * Values are prefixed with type: S: for string, N: for number, 0: for null
     */
    private fun parseValue(value: String): Any? {
        return when {
            value.startsWith("S:") -> value.substring(2)
            value.startsWith("N:") -> value.substring(2).toLongOrNull() ?: 0L
            value.startsWith("0:") -> null
            value == "null" -> null
            else -> value // Fallback to raw value
        }
    }

    // ========== Account Operations ==========

    private fun ensureAccountExists(id: String) {
        val existing = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertAccount(
                id = id,
                name = "",
                offbudget = 0,
                closed = 0,
                sort_order = null,
                tombstone = 0
            )
        }
    }

    private fun applyAccountColumn(id: String, column: String, value: Any?) {
        when (column) {
            "name" -> {
                val current = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
                if (current != null) {
                    db.actualDatabaseQueries.insertAccount(
                        id = id,
                        name = value as? String ?: "",
                        offbudget = current.offbudget,
                        closed = current.closed,
                        sort_order = current.sort_order,
                        tombstone = current.tombstone
                    )
                }
            }
            "offbudget" -> {
                val current = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
                if (current != null) {
                    db.actualDatabaseQueries.insertAccount(
                        id = id,
                        name = current.name,
                        offbudget = (value as? Long) ?: 0L,
                        closed = current.closed,
                        sort_order = current.sort_order,
                        tombstone = current.tombstone
                    )
                }
            }
            "closed" -> {
                val current = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
                if (current != null) {
                    db.actualDatabaseQueries.insertAccount(
                        id = id,
                        name = current.name,
                        offbudget = current.offbudget,
                        closed = (value as? Long) ?: 0L,
                        sort_order = current.sort_order,
                        tombstone = current.tombstone
                    )
                }
            }
            "sort_order" -> {
                val current = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
                if (current != null) {
                    db.actualDatabaseQueries.insertAccount(
                        id = id,
                        name = current.name,
                        offbudget = current.offbudget,
                        closed = current.closed,
                        sort_order = (value as? Long)?.toDouble(),
                        tombstone = current.tombstone
                    )
                }
            }
            "tombstone" -> {
                val current = db.actualDatabaseQueries.getAccountById(id).executeAsOneOrNull()
                if (current != null) {
                    db.actualDatabaseQueries.insertAccount(
                        id = id,
                        name = current.name,
                        offbudget = current.offbudget,
                        closed = current.closed,
                        sort_order = current.sort_order,
                        tombstone = (value as? Long) ?: 0L
                    )
                }
            }
        }
    }

    // ========== Payee Operations ==========

    private fun ensurePayeeExists(id: String) {
        val existing = db.actualDatabaseQueries.getPayeeById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertPayee(
                id = id,
                name = "",
                category = null,
                tombstone = 0,
                transfer_acct = null
            )
        }
    }

    private fun applyPayeeColumn(id: String, column: String, value: Any?) {
        val current = db.actualDatabaseQueries.getPayeeById(id).executeAsOneOrNull() ?: return
        when (column) {
            "name" -> db.actualDatabaseQueries.insertPayee(id, value as? String ?: "", current.category, current.tombstone, current.transfer_acct)
            "category" -> db.actualDatabaseQueries.insertPayee(id, current.name, value as? String, current.tombstone, current.transfer_acct)
            "tombstone" -> db.actualDatabaseQueries.insertPayee(id, current.name, current.category, (value as? Long) ?: 0L, current.transfer_acct)
            "transfer_acct" -> db.actualDatabaseQueries.insertPayee(id, current.name, current.category, current.tombstone, value as? String)
        }
    }

    // ========== Payee Mapping Operations ==========
    // NOTE: payee_mapping table has only 'id' and 'targetId' columns (no tombstone)

    private fun ensurePayeeMappingExists(id: String) {
        val existing = db.actualDatabaseQueries.getPayeeMappingById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertPayeeMapping(
                id = id,
                targetId = id  // Default: points to itself
            )
        }
    }

    private fun applyPayeeMappingColumn(id: String, column: String, value: Any?) {
        when (column) {
            "targetId" -> db.actualDatabaseQueries.insertPayeeMapping(id, value as? String ?: id)
        }
    }

    // ========== Category Operations ==========

    private fun ensureCategoryExists(id: String) {
        val existing = db.actualDatabaseQueries.getCategoryById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertCategory(
                id = id,
                name = "",
                cat_group = null,
                is_income = 0,
                sort_order = null,
                hidden = 0,
                tombstone = 0
            )
        }
    }

    private fun applyCategoryColumn(id: String, column: String, value: Any?) {
        val current = db.actualDatabaseQueries.getCategoryById(id).executeAsOneOrNull() ?: return
        when (column) {
            "name" -> db.actualDatabaseQueries.insertCategory(id, value as? String ?: "", current.cat_group, current.is_income, current.sort_order, current.hidden, current.tombstone)
            "cat_group" -> db.actualDatabaseQueries.insertCategory(id, current.name, value as? String, current.is_income, current.sort_order, current.hidden, current.tombstone)
            "is_income" -> db.actualDatabaseQueries.insertCategory(id, current.name, current.cat_group, (value as? Long) ?: 0L, current.sort_order, current.hidden, current.tombstone)
            "sort_order" -> db.actualDatabaseQueries.insertCategory(id, current.name, current.cat_group, current.is_income, (value as? Long)?.toDouble(), current.hidden, current.tombstone)
            "hidden" -> db.actualDatabaseQueries.insertCategory(id, current.name, current.cat_group, current.is_income, current.sort_order, (value as? Long) ?: 0L, current.tombstone)
            "tombstone" -> db.actualDatabaseQueries.insertCategory(id, current.name, current.cat_group, current.is_income, current.sort_order, current.hidden, (value as? Long) ?: 0L)
        }
    }

    // ========== Category Group Operations ==========

    private fun ensureCategoryGroupExists(id: String) {
        val existing = db.actualDatabaseQueries.getCategoryGroupById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertCategoryGroup(
                id = id,
                name = "",
                is_income = 0,
                sort_order = null,
                hidden = 0,
                tombstone = 0
            )
        }
    }

    private fun applyCategoryGroupColumn(id: String, column: String, value: Any?) {
        val current = db.actualDatabaseQueries.getCategoryGroupById(id).executeAsOneOrNull() ?: return
        when (column) {
            "name" -> db.actualDatabaseQueries.insertCategoryGroup(id, value as? String ?: "", current.is_income, current.sort_order, current.hidden, current.tombstone)
            "is_income" -> db.actualDatabaseQueries.insertCategoryGroup(id, current.name, (value as? Long) ?: 0L, current.sort_order, current.hidden, current.tombstone)
            "sort_order" -> db.actualDatabaseQueries.insertCategoryGroup(id, current.name, current.is_income, (value as? Long)?.toDouble(), current.hidden, current.tombstone)
            "hidden" -> db.actualDatabaseQueries.insertCategoryGroup(id, current.name, current.is_income, current.sort_order, (value as? Long) ?: 0L, current.tombstone)
            "tombstone" -> db.actualDatabaseQueries.insertCategoryGroup(id, current.name, current.is_income, current.sort_order, current.hidden, (value as? Long) ?: 0L)
        }
    }

    // ========== Transaction Operations ==========

    private fun ensureTransactionExists(id: String) {
        val existing = db.actualDatabaseQueries.getTransactionById(id).executeAsOneOrNull()
        if (existing == null) {
            db.actualDatabaseQueries.insertTransaction(
                id = id,
                acct = null,
                category = null,
                amount = 0,
                description = null,
                notes = null,
                date = null,
                sort_order = null,
                tombstone = 0,
                cleared = 1,
                pending = 0,
                reconciled = 0,
                isParent = 0,
                isChild = 0,
                parent_id = null,
                transferred_id = null
            )
        }
    }

    private fun applyTransactionColumn(id: String, column: String, value: Any?) {
        val current = db.actualDatabaseQueries.getTransactionById(id).executeAsOneOrNull() ?: return

        // Build updated transaction (simplified schema)
        // Note: 'description' column stores the payee ID
        val updated = when (column) {
            "acct" -> current.copy(acct = value as? String)
            "account" -> current.copy(acct = value as? String) // alias
            "category" -> current.copy(category = value as? String)
            "amount" -> current.copy(amount = (value as? Long))
            "description", "payee" -> current.copy(description = value as? String)
            "notes" -> current.copy(notes = value as? String)
            "date" -> current.copy(date = (value as? Long))
            "cleared" -> current.copy(cleared = (value as? Long))
            "pending" -> current.copy(pending = (value as? Long))
            "reconciled" -> current.copy(reconciled = (value as? Long))
            "sort_order" -> current.copy(sort_order = (value as? Long)?.toDouble())
            "tombstone" -> current.copy(tombstone = (value as? Long))
            // Split transaction columns
            "isParent", "is_parent" -> current.copy(isParent = (value as? Long))
            "isChild", "is_child" -> current.copy(isChild = (value as? Long))
            "parent_id" -> current.copy(parent_id = value as? String)
            // Transfer transaction column
            "transferred_id" -> current.copy(transferred_id = value as? String)
            // Ignore columns that don't exist in minimal schema
            else -> current
        }

        db.actualDatabaseQueries.insertTransaction(
            id = updated.id,
            acct = updated.acct,
            category = updated.category,
            amount = updated.amount ?: 0,
            description = updated.description,
            notes = updated.notes,
            date = updated.date,
            sort_order = updated.sort_order,
            tombstone = updated.tombstone ?: 0,
            cleared = updated.cleared ?: 1,
            pending = updated.pending ?: 0,
            reconciled = updated.reconciled ?: 0,
            isParent = updated.isParent ?: 0,
            isChild = updated.isChild ?: 0,
            parent_id = updated.parent_id,
            transferred_id = updated.transferred_id
        )
    }

    // ========== Budget Operations ==========

    private fun ensureBudgetExists(id: String) {
        val existing = db.actualDatabaseQueries.getBudgetById(id).executeAsOneOrNull()
        if (existing == null) {
            // Parse month and category from id (format: "YYYYMM-categoryId")
            val parts = id.split("-", limit = 2)
            val month = parts.getOrNull(0)?.toLongOrNull() ?: 0L
            val category = parts.getOrNull(1) ?: ""

            db.actualDatabaseQueries.insertBudget(
                id = id,
                month = month,
                category = category,
                amount = 0,
                carryover = 0,
                goal = null,
                tombstone = 0
            )
        }
    }

    private fun applyBudgetColumn(id: String, column: String, value: Any?) {
        val current = db.actualDatabaseQueries.getBudgetById(id).executeAsOneOrNull() ?: return

        when (column) {
            "month" -> db.actualDatabaseQueries.insertBudget(
                id, (value as? Long) ?: current.month, current.category,
                current.amount, current.carryover, current.goal, current.tombstone
            )
            "category" -> db.actualDatabaseQueries.insertBudget(
                id, current.month, (value as? String) ?: current.category,
                current.amount, current.carryover, current.goal, current.tombstone
            )
            "amount" -> db.actualDatabaseQueries.insertBudget(
                id, current.month, current.category,
                (value as? Long) ?: 0L, current.carryover, current.goal, current.tombstone
            )
            "carryover" -> db.actualDatabaseQueries.insertBudget(
                id, current.month, current.category,
                current.amount, (value as? Long) ?: 0L, current.goal, current.tombstone
            )
            "goal" -> db.actualDatabaseQueries.insertBudget(
                id, current.month, current.category,
                current.amount, current.carryover, value as? Long, current.tombstone
            )
            "tombstone" -> db.actualDatabaseQueries.insertBudget(
                id, current.month, current.category,
                current.amount, current.carryover, current.goal, (value as? Long) ?: 0L
            )
        }
    }

    // ========== Query Methods ==========

    fun getAccounts() = db.actualDatabaseQueries.getAccounts().executeAsList()

    fun getPayees() = db.actualDatabaseQueries.getPayees().executeAsList()

    fun getCategories() = db.actualDatabaseQueries.getCategories().executeAsList()

    fun getCategoryGroups() = db.actualDatabaseQueries.getCategoryGroups().executeAsList()

    fun getTransactionsByAccount(accountId: String) =
        db.actualDatabaseQueries.getTransactionsByAccount(accountId).executeAsList()

    fun getTransactionsByDateRange(startDate: Long, endDate: Long) =
        db.actualDatabaseQueries.getTransactionsByDateRange(startDate, endDate).executeAsList()

    fun getBudgetForMonth(month: Long) =
        db.actualDatabaseQueries.getBudgetForMonth(month).executeAsList()

    fun getBudgetForCategory(category: String) =
        db.actualDatabaseQueries.getBudgetForCategory(category).executeAsList()

    /**
     * Get spending totals by category for a date range.
     * Only includes transactions from on-budget accounts.
     * @param startDate Start date in YYYYMMDD format
     * @param endDate End date in YYYYMMDD format
     * @return List of (category, spent) pairs where spent is negative for expenses
     */
    fun getSpentByCategory(startDate: Long, endDate: Long) =
        db.actualDatabaseQueries.getSpentByCategory(startDate, endDate).executeAsList()

    /**
     * Get categories with their group information pre-joined.
     * Useful for budget views that need to display categories grouped.
     * Only returns visible (non-hidden, non-tombstoned) categories and groups.
     */
    fun getCategoriesWithGroups() =
        db.actualDatabaseQueries.getCategoriesWithGroups().executeAsList()

    /**
     * Get the balance for a specific account.
     * @param accountId The account ID
     * @return The sum of all transaction amounts for this account
     */
    fun getAccountBalance(accountId: String): Long =
        db.actualDatabaseQueries.getAccountBalance(accountId).executeAsOne().toLong()

    fun getLastSyncTimestamp(): String? =
        db.actualDatabaseQueries.getLastTimestamp().executeAsOneOrNull()?.last_ts

    fun setSyncMetadata(key: String, value: String) =
        db.actualDatabaseQueries.setSyncMetadata(key, value)

    fun getSyncMetadata(key: String): String? =
        db.actualDatabaseQueries.getSyncMetadata(key).executeAsOneOrNull()?.value_

    /**
     * Clear all data from the database.
     */
    fun clearAll() {
        db.actualDatabaseQueries.clearMessages()
        db.actualDatabaseQueries.clearTransactions()
        db.actualDatabaseQueries.clearCategories()
        db.actualDatabaseQueries.clearCategoryGroups()
        db.actualDatabaseQueries.clearPayees()
        db.actualDatabaseQueries.clearAccounts()
        db.actualDatabaseQueries.clearBudgets()
        db.actualDatabaseQueries.clearMetadata()
    }
}

File: dart/lib/src/db/sync_repository.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:drift/drift.dart';

import '../proto/sync_messages.dart' as proto;
import 'database.dart';

/// Repository for syncing CRDT messages to the local database.
class SyncRepository {
  final ActualDatabase db;

  SyncRepository(this.db);

  /// Apply a list of message envelopes to the database.
  /// Messages are stored in the CRDT log and applied to their respective tables.
  Future<void> applyMessages(List<proto.MessageEnvelope> envelopes) async {
    for (final envelope in envelopes) {
      if (!envelope.isEncrypted) {
        try {
          final message = envelope.decodeMessage();
          await applyMessage(envelope.timestamp, message);
        } catch (e) {
          // Log error but continue processing
          print('Failed to apply message ${envelope.timestamp}: $e');
        }
      }
    }
  }

  /// Apply a single CRDT message.
  Future<void> applyMessage(String timestamp, proto.Message message) async {
    // Store in CRDT log (value stored as BLOB)
    await db.insertMessage(
      timestamp: timestamp,
      dataset: message.dataset,
      row: message.row,
      column: message.column,
      value: Uint8List.fromList(utf8.encode(message.value)),
    );

    // Parse the value
    final parsedValue = parseValue(message.value);

    // Ensure row exists, then update the column
    switch (message.dataset) {
      case 'accounts':
        await _ensureAccountExists(message.row);
        await _applyAccountColumn(message.row, message.column, parsedValue);
        break;
      case 'payees':
        await _ensurePayeeExists(message.row);
        await _applyPayeeColumn(message.row, message.column, parsedValue);
        break;
      case 'payee_mapping':
        // Handle payee mapping
        break;
      case 'categories':
        await _ensureCategoryExists(message.row);
        await _applyCategoryColumn(message.row, message.column, parsedValue);
        break;
      case 'category_groups':
        await _ensureCategoryGroupExists(message.row);
        await _applyCategoryGroupColumn(
            message.row, message.column, parsedValue);
        break;
      case 'transactions':
        await _ensureTransactionExists(message.row);
        await _applyTransactionColumn(
            message.row, message.column, parsedValue);
        break;
      case 'zero_budgets':
      case 'zero_budget_months':
        // Handle budget data
        break;
      case 'schedules':
        // Handle schedules
        break;
      case 'rules':
        // Handle rules
        break;
      case 'notes':
        // Handle notes
        break;
      default:
        // Unknown dataset, skip
        break;
    }
  }

  /// Parse Actual's typed value format.
  /// Values are prefixed with type: S: for string, N: for number, 0: for null
  Object? parseValue(String value) {
    if (value.startsWith('S:')) {
      return value.substring(2);
    } else if (value.startsWith('N:')) {
      return int.tryParse(value.substring(2)) ?? 0;
    } else if (value.startsWith('0:') || value == 'null') {
      return null;
    } else {
      return value; // Fallback to raw value
    }
  }

  /// Encode a value to Actual's typed format.
  static String encodeValue(Object? value) {
    if (value == null) {
      return '0:';
    } else if (value is int) {
      return 'N:$value';
    } else if (value is double) {
      return 'N:${value.toInt()}';
    } else {
      return 'S:$value';
    }
  }

  // ========== Account Operations ==========

  Future<void> _ensureAccountExists(String id) async {
    final existing = await db.getAccountById(id);
    if (existing == null) {
      await db.insertAccount(AccountsCompanion(
        id: Value(id),
        name: const Value(''),
        offbudget: const Value(0),
        closed: const Value(0),
        sortOrder: const Value(null),
        tombstone: const Value(0),
      ));
    }
  }

  Future<void> _applyAccountColumn(
      String id, String column, Object? value) async {
    final current = await db.getAccountById(id);
    if (current == null) return;

    await db.insertAccount(AccountsCompanion(
      id: Value(id),
      name: Value(column == 'name' ? (value as String?) ?? '' : current.name),
      offbudget: Value(
          column == 'offbudget' ? (value as int?) ?? 0 : current.offbudget),
      closed:
          Value(column == 'closed' ? (value as int?) ?? 0 : current.closed),
      sortOrder: Value(column == 'sort_order'
          ? (value as int?)?.toDouble()
          : current.sortOrder),
      tombstone: Value(
          column == 'tombstone' ? (value as int?) ?? 0 : current.tombstone),
    ));
  }

  // ========== Payee Operations ==========

  Future<void> _ensurePayeeExists(String id) async {
    final existing = await db.getPayeeById(id);
    if (existing == null) {
      await db.insertPayee(PayeesCompanion(
        id: Value(id),
        name: const Value(''),
        category: const Value(null),
        tombstone: const Value(0),
      ));
    }
  }

  Future<void> _applyPayeeColumn(
      String id, String column, Object? value) async {
    final current = await db.getPayeeById(id);
    if (current == null) return;

    await db.insertPayee(PayeesCompanion(
      id: Value(id),
      name: Value(column == 'name' ? (value as String?) ?? '' : current.name),
      category: Value(
          column == 'category' ? value as String? : current.category),
      tombstone: Value(
          column == 'tombstone' ? (value as int?) ?? 0 : current.tombstone),
    ));
  }

  // ========== Category Operations ==========

  Future<void> _ensureCategoryExists(String id) async {
    final existing = await db.getCategoryById(id);
    if (existing == null) {
      await db.insertCategory(CategoriesCompanion(
        id: Value(id),
        name: const Value(''),
        catGroup: const Value(null),
        isIncome: const Value(0),
        sortOrder: const Value(null),
        hidden: const Value(0),
        tombstone: const Value(0),
      ));
    }
  }

  Future<void> _applyCategoryColumn(
      String id, String column, Object? value) async {
    final current = await db.getCategoryById(id);
    if (current == null) return;

    await db.insertCategory(CategoriesCompanion(
      id: Value(id),
      name: Value(column == 'name' ? (value as String?) ?? '' : current.name),
      catGroup: Value(
          column == 'cat_group' ? value as String? : current.catGroup),
      isIncome: Value(
          column == 'is_income' ? (value as int?) ?? 0 : current.isIncome),
      sortOrder: Value(column == 'sort_order'
          ? (value as int?)?.toDouble()
          : current.sortOrder),
      hidden:
          Value(column == 'hidden' ? (value as int?) ?? 0 : current.hidden),
      tombstone: Value(
          column == 'tombstone' ? (value as int?) ?? 0 : current.tombstone),
    ));
  }

  // ========== Category Group Operations ==========

  Future<void> _ensureCategoryGroupExists(String id) async {
    final existing = await db.getCategoryGroupById(id);
    if (existing == null) {
      await db.insertCategoryGroup(CategoryGroupsCompanion(
        id: Value(id),
        name: const Value(''),
        isIncome: const Value(0),
        sortOrder: const Value(null),
        hidden: const Value(0),
        tombstone: const Value(0),
      ));
    }
  }

  Future<void> _applyCategoryGroupColumn(
      String id, String column, Object? value) async {
    final current = await db.getCategoryGroupById(id);
    if (current == null) return;

    await db.insertCategoryGroup(CategoryGroupsCompanion(
      id: Value(id),
      name: Value(column == 'name' ? (value as String?) ?? '' : current.name),
      isIncome: Value(
          column == 'is_income' ? (value as int?) ?? 0 : current.isIncome),
      sortOrder: Value(column == 'sort_order'
          ? (value as int?)?.toDouble()
          : current.sortOrder),
      hidden:
          Value(column == 'hidden' ? (value as int?) ?? 0 : current.hidden),
      tombstone: Value(
          column == 'tombstone' ? (value as int?) ?? 0 : current.tombstone),
    ));
  }

  // ========== Transaction Operations ==========

  Future<void> _ensureTransactionExists(String id) async {
    final existing = await db.getTransactionById(id);
    if (existing == null) {
      await db.insertTransaction(TransactionsCompanion(
        id: Value(id),
        acct: const Value(null),
        category: const Value(null),
        amount: const Value(0),
        description: const Value(null),
        notes: const Value(null),
        date: const Value(null),
        sortOrder: const Value(null),
        tombstone: const Value(0),
        cleared: const Value(1),
      ));
    }
  }

  Future<void> _applyTransactionColumn(
      String id, String column, Object? value) async {
    final current = await db.getTransactionById(id);
    if (current == null) return;

    // Build updated transaction
    // Note: 'description' column stores the payee ID
    final acct = (column == 'acct' || column == 'account')
        ? value as String?
        : current.acct;
    final category =
        column == 'category' ? value as String? : current.category;
    final amount = column == 'amount' ? (value as int?) : current.amount;
    final description = (column == 'description' || column == 'payee')
        ? value as String?
        : current.description;
    final notes = column == 'notes' ? value as String? : current.notes;
    final date = column == 'date' ? (value as int?) : current.date;
    final cleared = column == 'cleared' ? (value as int?) : current.cleared;
    final sortOrder = column == 'sort_order'
        ? (value as int?)?.toDouble()
        : current.sortOrder;
    final tombstone =
        column == 'tombstone' ? (value as int?) : current.tombstone;

    await db.insertTransaction(TransactionsCompanion(
      id: Value(id),
      acct: Value(acct),
      category: Value(category),
      amount: Value(amount ?? 0),
      description: Value(description),
      notes: Value(notes),
      date: Value(date),
      sortOrder: Value(sortOrder),
      tombstone: Value(tombstone ?? 0),
      cleared: Value(cleared ?? 1),
    ));
  }

  // ========== Query Methods ==========

  Future<List<Account>> getAccounts() => db.getAccounts();

  Future<List<Payee>> getPayees() => db.getPayees();

  Future<List<Category>> getCategories() => db.getCategories();

  Future<List<CategoryGroup>> getCategoryGroups() => db.getCategoryGroups();

  Future<List<Transaction>> getTransactionsByAccount(String accountId) =>
      db.getTransactionsByAccount(accountId);

  Future<List<Transaction>> getTransactionsByDateRange(
          int startDate, int endDate) =>
      db.getTransactionsByDateRange(startDate, endDate);

  Future<String?> getLastSyncTimestamp() => db.getLastTimestamp();

  Future<void> setSyncMetadata(String key, String value) =>
      db.setSyncMetadataValue(key, value);

  Future<String?> getSyncMetadata(String key) =>
      db.getSyncMetadataValue(key);

  /// Clear all data from the database.
  Future<void> clearAll() => db.clearAllData();
}
